src/QsCompiler/Transformations/CallGraph/SimpleCallGraph.cs

```diff
@@ -0,0 +1,465 @@
+        /// </summary>
+        public SimpleCallGraph(QsCompilation compilation, bool trim = false)
+        {
+            if (compilation is null)
+            {
+                throw new ArgumentNullException(nameof(compilation));
+            }
+
+            if (trim)
+            {
+                BuildCallGraph.PopulateTrimmedGraph(this.graphBuilder, compilation);
+            }
+            else
+            {
+                BuildCallGraph.PopulateSimpleGraph(this.graphBuilder, compilation);
+            }
+        }
+
+        /// <summary>
+        /// Constructs a call graph from callables.
+        /// Throws an ArgumentNullException if the argument is null or any of the callables are null.
+        /// </summary>
+        public SimpleCallGraph(IEnumerable<QsCallable> callables)
+        {
+            if (callables is null || callables.Any(x => x is null))
+            {
+                throw new ArgumentNullException(nameof(callables));
+            }
+
+            BuildCallGraph.PopulateSimpleGraph(this.graphBuilder, callables);
+        }
+
+        // Member Methods
+
+        /// <summary>
+        /// Returns the children nodes of a given node. Each key in the returned lookup is a child
+        /// node of the given node. Each value in the lookup is an edge connecting the given node to
+        /// the child node represented by the associated key.
+        /// Returns an empty ILookup if the node was found with no dependencies or was not found in
+        /// the graph.
+        /// Throws ArgumentNullException if argument is null.
+        /// </summary>
+        public ILookup<SimpleCallGraphNode, SimpleCallGraphEdge> GetDirectDependencies(SimpleCallGraphNode node) => this.graphBuilder.GetDirectDependencies(node);
+
+        /// <summary>
+        /// Given a call graph edges, finds all cycles and determines if each is valid.
+        /// Invalid cycles are those that cause type parameters to be mapped to
+        /// other type parameters of the same callable (constricting resolutions)
+        /// or to a type containing a nested reference to the same type parameter,
+        /// i.e Foo.A -> Foo.A[].
+        /// Returns an enumerable of tuples for each edge of each invalid cycle found,
+        /// each tuple containing a diagnostic and the callable name where the diagnostic
+        /// should be placed.
+        /// </summary>
+        public IEnumerable<(QsCompilerDiagnostic, QsQualifiedName)> VerifyAllCycles()
+        {
+            var diagnostics = new List<(QsCompilerDiagnostic, QsQualifiedName)>();
+
+            if (this.Nodes.Any())
+            {
+                var cycles = this.GetCallCycles().SelectMany(x => CartesianProduct(this.GetEdges(x).Reverse()));
+                foreach (var cycle in cycles)
+                {
+                    var combination = new TypeResolutionCombination(cycle.Select(edge => edge.ParamResolutions).ToArray());
```

Comment:
It's a little awkward that you need to call `ToArray()`, but `TypeResolutionCombination` calls `Where` which only requires an `IEnumerable`. Maybe `TypeResolutionCombination` could take an `IEnumerable` instead?

Reply:
done